### PR TITLE
feat(desktop): tray startup UX — orphan cloudflared cleanup, surfaced logs, tunnel warming state (#2835, #2836)

### LINKED ISSUE #2835: bug(desktop): Tauri tray reports "Server failed to start / Health check timeout after 60s" when the bundled server is actually running fine
## Symptom
Tauri tray app shows a loading screen with red error text:
- First attempt: **"Server failed to start within 60 seconds."**
- Subsequent attempts: **"Health check timeout after 60s."**

Despite the error, in diagnostics the server is actually running fine:
- node child is spawned successfully
- WS listening on `0.0.0.0:8765` within ~1s
- `curl http://127.0.0.1:8765/` returns **200 OK** in <1ms
- Client sockets (including the dashboard) connect and authenticate
- Cloudflare tunnel establishes normally

## Repro
1. Launch Chroxy.app
2. Quit or crash while `cloudflared` is still running
3. Relaunch — hangs for 60s then shows the error
4. Repeat attempts may alternate between the outer timeout and the inner "Health check timeout"

## Diagnostics (captured via NODE_OPTIONS logger hook)
On a fresh launch where the tray showed a 60s timeout, the child's own log shows a clean start:
\`\`\`
[stdout] ║ Chroxy Server v0.6.10 (claude-sdk (SDK mode)) ║
[stdout] [ws] Server listening on 0.0.0.0:8765 (cli mode)
[stdout] [ws] Client 0020e5bb connected (awaiting auth)
[stdout] [ws] Client 0020e5bb authenticated
[stdout] [tunnel] Cloudflare tunnel established: ...
\`\`\`
The client `0020e5bb` is the dashboard itself — yet the Rust outer layer still reports failure.

## Likely causes
1. **Orphan `cloudflared` processes** survive a crashed Chroxy. `ServerManager::start()` calls `kill_port_holder(port)` but doesn't touch cloudflared. The orphan may interfere with tunnel setup or with local networking state.
2. **Race in `start_health_poll`** at `packages/desktop/src-tauri/src/server.rs:493-568`. The 60s loop uses 2s intervals with 2s ureq timeout. Under macOS GUI sandbox networking, something's silently preventing success despite curl working.
3. **Silent poll thread death** — if ureq panics or the generation counter advances unexpectedly, the thread returns without ever flipping status to Running.

## Fixes to consider
1. **Kill orphan `cloudflared` on start** — in `start_server_process`, after `kill_port_holder`, also pkill `cloudflared tunnel --url http://localhost:<port>`.
2. **Surface the captured server logs in the error screen.** Rust already has `log_buffer` (last 100 lines). On timeout, render the last ~30 lines in the error view with a "Copy logs" button so users can self-diagnose.
3. **Instrument `start_health_poll`** — log every ureq attempt/result (currently silent) so we can see which requests fail and why.
4. **Health-poll startup probe** — before the 2s loop, do one immediate probe; if the bind is already up the user gets instant feedback.

## Files
- `packages/desktop/src-tauri/src/server.rs:229-371` (start_server_process)
- `packages/desktop/src-tauri/src/server.rs:493-568` (start_health_poll)
- `packages/desktop/src-tauri/src/lib.rs:788-829` (monitor_startup, outer 60s)

## Workaround
\`\`\`
pkill -f cloudflared
pkill -f chroxy-desktop
open /Applications/Chroxy.app
\`\`\`

---

### What's *not* broken (verified, no reinstall needed)
- **`.app` bundle integrity** — `/Applications/Chroxy.app/Contents/Resources/server/src/cli.js` runs cleanly when invoked manually with Node 22
- **Node 22 auto-discovery** — `resolve_node22` in `packages/desktop/src-tauri/src/node.rs` finds `/opt/homebrew/opt/node@22/bin/node` correctly
- **macOS entitlements** — `entitlements.plist` has the expected Hardened Runtime exceptions (JIT, library validation disabled, etc.); spawning is not being blocked
- **Code signing** — present and valid

A user seeing this error does **not** need to reinstall Chroxy.app. That's a red herring.

### Tunnel routability phase is expected but ambiguous (possible split-out)
After the Cloudflare tunnel is created, the server runs an internal routability check (`tunnel-check.js`) that retries up to 20 times with backoff — cold Quick Tunnels legitimately take 30–90 seconds to become DNS-propagated. The server log during this phase is:
```
[tunnel] Cloudflare tunnel established: HTTP=https://...trycloudflare.com
[tunnel] Verifying tunnel is routable...
[tunnel] Attempt 1/20 failed: fetch failed
[tunnel] Attempt 2/20 failed: fetch failed
...
```
This is **normal** and should not be conflated with the local 60s startup window. But if the Rust health poll flips Running quickly (as it should — local 127.0.0.1:8765 is up in <1s), the user may still see the tray dashboard before the tunnel is routable, with no indication that the tunnel is "still warming up". Consider a separate issue for: **"Distinguish 'local server up, tunnel warming' from 'failed' in the tray UI during the first ~90s."**

### Reproducible diagnostic technique
The Rust layer pipes the child's stdout/stderr into an in-memory `log_buffer` that's only visible via the tray's Console menu item — so when the tray shows the generic 60s timeout but the user hasn't clicked Console yet, the logs are lost. To capture the child's output externally during triage:

```bash
cat > /tmp/chroxy-logger.mjs <<'JS'
import fs from 'fs'
const LOG = '/tmp/chroxy-child.log'
const log = (...a) => { try { fs.appendFileSync(LOG, a.join(' ') + '\n') } catch {} }
log('=== Child started', new Date().toISOString(), '===')
log('argv:', process.argv.join(' '))
const origOut = process.stdout.write.bind(process.stdout)
const origErr = process.stderr.write.bind(process.stderr)
process.stdout.write = (c, ...r) => { try { fs.appendFileSync(LOG, '[stdout] ' + (Buffer.isBuffer(c) ? c.toString() : String(c))) } catch {}; return origOut(c, ...r) }
process.stderr.write = (c, ...r) => { try { fs.appendFileSync(LOG, '[stderr] ' + (Buffer.isBuffer(c) ? c.toString() : String(c))) } catch {}; return origErr(c, ...r) }
process.on('exit', code => log('=== Exit code', code, '==='))
process.on('uncaughtException', e => log('UNCAUGHT:', e.stack || String(e)))
JS

NODE_OPTIONS='--import file:///tmp/chroxy-logger.mjs' \
  /Applications/Chroxy.app/Contents/MacOS/chroxy-desktop
```

Then inspect `/tmp/chroxy-child.log`. NODE_OPTIONS is inherited by the spawned child through the Rust process's env (the Rust spawn uses `Stdio::piped()` but does not strip NODE_OPTIONS). This is how the diagnostic output in the "Diagnostics" section above was captured.

Once the in-tray "show server logs on error" fix lands, this external workaround can be retired.

### LINKED ISSUE #2836: bug(desktop): tray UI does not distinguish "local server up, tunnel warming" from "failed" during first ~90s
## Context
Split out from #2835 (which focuses on the false-timeout root cause). This issue is about the **separate UX gap** that persists even when server startup is healthy.

## Symptom
When Chroxy starts a Quick Tunnel, the local server binds port 8765 within ~1s, but the Cloudflare tunnel URL takes 30–90s to become DNS-routable (a Cloudflare-side propagation delay, not our bug). During that window:

1. Rust health poll flips to Running quickly (local `127.0.0.1:8765` is up)
2. Tauri window shows the dashboard or a generic loading state
3. The server logs `[tunnel] Attempt N/20 failed: fetch failed` repeatedly until propagation completes
4. The user sees **nothing** that communicates "tunnel is warming up, QR/URL will appear shortly"

From the user's perspective, this looks indistinguishable from a broken tunnel, even though it's the normal first-~90s of a Quick Tunnel's life.

## What the UI should do
Add a distinct phase between "local server up" and "tunnel routable":

- **State:** `tunnel_warming` (or extend the existing server-status enum)
- **Visual:** a small banner/toast: "Tunnel warming up… (QR will appear shortly)"
- **Progress:** show attempt count from server log (e.g. "Verifying tunnel… attempt 3/20")
- **Failure transition:** only after all 20 attempts fail (or cloudflared exits) should we surface a real error

## Signal source
The server already broadcasts `server_status` events during the routability loop — see `packages/server/src/tunnel/` and the `tunnel-check.js` utility. The tray/dashboard just needs to render that state instead of treating it as indeterminate.

## Files
- `packages/server/src/tunnel/base.js` / `cloudflare.js` / `tunnel-check.js` — verification loop + status emits
- `packages/dashboard/src/store/` — consume `server_status` events
- `packages/desktop/src-tauri/src/window.rs` (loading page) or the React loading screen

## Severity
Medium — not a correctness bug (the tunnel will come up), but the current behavior reads as "broken" to first-time users and drives false recovery attempts (quit, reinstall, pkill).

## Relationship to other issues
- Related: #2835 (the false 60s timeout, which shares the underlying "we don't show server logs / tunnel phase to the user" theme)
- Once the log-surfacing fix in #2835 lands, this issue's fix can reuse the same channel

### PR DESCRIPTION
**Pull request by blamechris**
## Summary

Bundled fix for two related tray startup UX issues.

### Issue #2835 — "Server failed to start" when it's actually running
- **Auto-kill orphan cloudflared.** `kill_orphan_cloudflared(port)` runs alongside `kill_port_holder(port)` in `start_server_process`. Uses `ps -eo pid=,command=` (unix) or `wmic process get` (windows) to enumerate, then a pure `cloudflared_pids_to_kill()` filter picks processes with cloudflared in argv AND a `http://{localhost,127.0.0.1}:{port}` URL at a word boundary (so 876 never matches 8765).
- **Log every health-poll attempt.** `start_health_poll` now `eprintln!`s `[health_poll] attempt #N GET <url> -> <status> (Nms)` per attempt and, on timeout, a one-line summary with totals (`N attempts: X non-200, Y errors`) embedded in the error message so #2832 can be reproduced from the log buffer alone.
- **Surface logs in error UI.** New `get_startup_logs(limit)` Tauri command + loading.js renders the last 30 server log lines in a collapsible `<details>` block with a "Copy logs" button whenever `server_error` fires.

### Issue #2836 — "local server up, tunnel warming" vs "failed"
- Server: `server-cli.js` now emits `phase: 'tunnel_warming'` (new wire name) on `server_status` broadcasts during `waitForTunnel`, carrying `attempt`, `maxAttempts`, `tunnelMode`, and `tunnelUrl`.
- Dashboard: message handler normalizes both `tunnel_warming` (current) and `tunnel_verifying` (legacy alias) to `serverPhase: 'tunnel_warming'` so in-flight clients on older servers stay working.
- New banner on App.tsx: "Tunnel warming up… attempt N/20 (QR will appear shortly)". Status bar and ConsolePage wording updated to match.
- Transitions to `ready` only after the server broadcasts `phase: 'ready'`; throws `TUNNEL_NOT_ROUTABLE` after 20 failed attempts (unchanged behavior).

## Test plan

- [x] Rust: `cargo test --manifest-path packages/desktop/src-tauri/Cargo.toml` → 75/75 pass (+10 new cloudflared filter tests).
- [x] Server: `npm --prefix packages/server test` → 3222/3227 pass (+4 new tunnel-warming-broadcast tests; 4 pre-existing flaky integration tests on this machine's `cloudflared` / WebTaskManager — same as baseline `main`).
- [x] Dashboard: `npm --prefix packages/dashboard test` → 1160/1160 (+9 new tests: 4 message-handler, 5 App banner).
- [x] Lint: `npm --prefix packages/server run lint` → 0 errors (4 pre-existing warnings).
- [x] Typecheck: `npm --prefix packages/dashboard run typecheck` → clean.
- [ ] Manual smoke: start the desktop app with a stale `cloudflared` holding port 8765 from a prior crash, verify it gets reaped and startup succeeds without "Server failed to start".
- [ ] Manual smoke: on a slow network, verify the "Tunnel warming up… N/20" banner appears within 2s of server becoming healthy and disappears when the QR loads.
- [ ] Manual smoke: force a startup failure (bad config / corrupt node binary), verify the error screen shows a "Show server logs (N lines)" expander with a working Copy button.

Closes #2835
Closes #2836

### CODE REVIEW

**Comment by blamechris** [comment]
## Code Review: PR #2845

### Summary

Bundled fix for #2835 (orphan cloudflared cleanup, health-poll logging, startup-error log surfacing) and #2836 (tunnel_warming phase + dashboard banner). 12 files, +859 / -127. Clean TDD: pure `cloudflared_pids_to_kill()` filter extracted for 10 Rust unit tests, new `tunnel-warming-broadcast.test.js` (+4), new dashboard message-handler tests (+4), App.tsx banner tests (+5). All targeted tests pass; CI green on main Server/App/Protocol/Type-Check jobs.

### Strengths

- **Word-boundary port discrimination** in `cloudflared_pids_to_kill` handles the 876-vs-8765 edge case correctly — the unit tests explicitly pin this.
- **Pure filter extraction** means orphan-cleanup logic is unit-testable without spawning anything. Rust test count jumped 65 → 75.
- **Health-poll counters** (`attempts`/`non200`/`network_errors`) embed the summary directly in the `ServerStatus::Error` message — survives into the user-visible error text, not just stderr.
- **Backward compat done right**: dashboard handler accepts both `tunnel_warming` (new) and `tunnel_verifying` (legacy) but always normalizes to `tunnel_warming`. Older in-flight clients keep working.
- **Protocol extension, not a new event**: stays within the "extend existing `server_status` event shape" constraint from the issue.
- **No attribution leaks** in commit message or PR body.

### Issues Found

#### Critical (Must Fix)

None.

#### Suggestions (Should Consider)

| File | Line | Suggestion | Rationale |
|------|------|------------|-----------|
| `packages/desktop/src-tauri/src/server.rs` | health_poll block | `eprintln!` writes go to the tray process's stderr, NOT to `self.log_buffer` (which only captures the server *child*'s stdout/stderr). The "Copy logs" button in the error UI from sub-fix C therefore won't include health-poll diagnostics. | Tracked as [#2846](https://github.com/blamechris/chroxy/issues/2846) — either tee into a second ring buffer or adopt a `log` crate sink. |
| `packages/dashboard/src/App.tsx` | new banner | Banner uses inline `style={{...}}` despite already having `className="tunnel-warming-banner"`. Hard-codes hex colors; won't participate in theme CSS variables. | Tracked as [#2847](https://github.com/blamechris/chroxy/issues/2847) — move rules into CSS class. |
| `packages/desktop/dist/loading.js` | `renderStartupLogs` | `if (document.getElementById('startup-logs')) return` short-circuit means a SECOND `server_error` (e.g. after user hits Retry) shows stale logs from the first failure. | Tracked as [#2848](https://github.com/blamechris/chroxy/issues/2848) — remove+re-render on repeated events. |
| `packages/dashboard/src/store/message-handler.ts` | server_status branch | Mixed-version case: new server emits `tunnel_warming`, old cached dashboard only knows `tunnel_verifying`; falls through to legacy branch and `tunnel_warming` messages land in the chat log as system messages. | Tracked as [#2849](https://github.com/blamechris/chroxy/issues/2849) — decide migration strategy (dual-emit for one release, or drop unknown phases silently). |

#### Nitpicks (Optional)

- Diff includes ~180 lines of unrelated `cargo fmt` reformats in `lib.rs` (argument wrapping, menu builder indentation, etc.). They're correct per rustfmt but bloat the review surface. Reverting or splitting into a separate fmt-only commit is an option for future PRs; not worth re-doing here.
- `get_startup_logs(limit: Option<usize>)` has no upper bound. Defensible since `log_buffer` is capped at 100, but clamping to e.g. 100 in the command would be more self-documenting.
- The `\u2026` (horizontal ellipsis) in "Tunnel warming up…" is correct UX polish but worth remembering if i18n ever appears — some terminals render it as `?`.

### Deferred Items (Follow-Up Issues)

| Suggestion | Issue | Rationale for deferral |
|------------|-------|------------------------|
| Surface tray-process stderr (incl. health-poll logs) in `get_startup_logs` | [#2846](https://github.com/blamechris/chroxy/issues/2846) | Requires a second buffer + `log` sink; non-trivial infra change, and the task explicitly says health-poll logging is "part of making #2832 reproducible later" — terminal-level inspection is the current target. |
| Move tunnel-warming banner inline styles to CSS class | [#2847](https://github.com/blamechris/chroxy/issues/2847) | Theme engine uplift is a separate concern; banner works correctly today. |
| Refresh startup logs on repeated `server_error` | [#2848](https://github.com/blamechris/chroxy/issues/2848) | Edge case (Retry flow); current behaviour is "stale logs visible" not "wrong logs". |
| Mixed-version dashboard vs new-server phase name | [#2849](https://github.com/blamechris/chroxy/issues/2849) | Impacts only users with aggressively cached SPA across a server upgrade; bundled install path is safe. |

### Architecture Notes

- `cloudflared_pids_to_kill` is a clean pure function in `server.rs` module scope — it doesn't need `ServerManager`. Good decision: keeps testability independent of `Mutex` / tauri state.
- `phase: 'tunnel_warming'` is a reasonable contract: it's a semantic status on an existing `server_status` event, so it doesn't need protocol-version gating.
- The `get_startup_logs` Tauri command is the right choice over stuffing logs into the event payload — logs can be up to 20KB and events fire frequently during startup.

### Verdict

- [x] **Approve** — ready to merge once the 4 follow-up issues are acknowledged. No blockers. The sub-fix B caveat (health-poll logs not in log_buffer) is the only meaningful functional gap and is tracked.

Smoke-test checkboxes in the PR body remain unchecked — merging should wait for at least the first two to be confirmed, but that's a pre-merge gate rather than a review block.

File: packages/desktop/src-tauri/src/server.rs [hunk 1]
```diff
@@ -200,12 +271,104 @@ impl ServerManager {
...
+            }
+            let pid = match parts[0].trim().parse::<u32>() {
+                Ok(p) => p,
+                Err(_) => continue,
+            };
+            let rest = parts[1];
+            // Drop the leading "Node," segment to get just CommandLine.
+            let cmd = rest.splitn(2, ',').nth(1).unwrap_or("").to_string();
+            procs.push((pid, cmd));
+        }
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/desktop/src-tauri/src/server.rs:367]: **FOLLOW-UP ISSUE**

Created https://github.com/blamechris/chroxy/issues/2850 to track this.

**Reason for deferral:** The suggested patch is ~100 lines of Windows-specific enumeration (wmic → PowerShell fallback + a proper CSV parser for quoted command lines with embedded commas + its own test surface). Chroxy's primary target is macOS per CLAUDE.md, and `kill_orphan_cloudflared` is a best-effort resilience feature — on newer Windows where wmic is removed, the Windows path is now a no-op rather than crashing, so there's no regression. Out of scope for this #2835/#2836 bundle; tracked to restore working Windows cleanup separately.

File: packages/desktop/src-tauri/src/server.rs [hunk 2]
```diff
@@ -11,6 +11,75 @@ use crate::config::{self, ChroxyConfig};
 use crate::lock_or_recover;
 use crate::node;
 
+/// Pure filter: given an iterator of (pid, full_command_line) pairs, return
+/// the pids whose command line matches a `cloudflared tunnel --url
+/// http(s)://{localhost|127.0.0.1}:{port}` invocation.
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/desktop/src-tauri/src/server.rs]: **FIX**

Fixed in `8bff12bd2`

**Change:** Rewrote the docstring to match actual behavior. cloudflared Quick/Named tunnels always point at the local server over plain HTTP (there's no TLS at the origin), so `https://` never appears in the needles — the comment was the outlier, not the code.

```diff
 /// Pure filter: given an iterator of (pid, full_command_line) pairs, return
 /// the pids whose command line matches a `cloudflared tunnel --url
-/// http(s)://{localhost|127.0.0.1}:{port}` invocation.
+/// http://{localhost|127.0.0.1}:{port}` invocation. (Cloudflare tunnels
+/// always point at the local server over plain HTTP — there's no https
+/// variant on this code path.)
```

File: packages/server/tests/tunnel-warming-broadcast.test.js [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+  return {
+    type: 'server_status',
+    phase: 'tunnel_warming',
+    tunnelMode,
+    tunnelUrl,
+    attempt,
+    maxAttempts,
+    message: `Tunnel warming up… (${attempt}/${maxAttempts})`,
+  }
+}
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/server/tests/tunnel-warming-broadcast.test.js]: **FIX**

Fixed in `8bff12bd2`

**Change:** Chose option (b) — extracted the broadcast construction into real production helpers (`buildTunnelWarmingStatus`, `buildTunnelReadyStatus`) in `server-cli.js`, and updated the test to import them directly. Tests now pin the same object that ships to clients, so a refactor at the broadcast site will fail the assertions instead of silently dropping fields.

Also added a field-whitelist assertion so accidentally adding a stray field to the helper will fail the test.

```diff
// server-cli.js (new)
+export function buildTunnelWarmingStatus({ tunnelMode, tunnelUrl, attempt, maxAttempts }) {
+  const base = { type: 'server_status', phase: 'tunnel_warming', tunnelMode, tunnelUrl }
+  if (typeof attempt === 'number' && typeof maxAttempts === 'number') {
+    return { ...base, attempt, maxAttempts, message: `Tunnel warming up… (${attempt}/${maxAttempts})` }
+  }
+  return { ...base, message: 'Tunnel warming up…' }
+}

// Call site — was 12-line inline object, now:
-      wsServer.broadcast({ type: 'server_status', phase: 'tunnel_warming', tunnelMode: tunnelArg.mode, tunnelUrl: httpUrl, attempt, maxAttempts, message: `Tunnel warming up… (${attempt}/${maxAttempts})` })
+      wsServer.broadcast(buildTunnelWarmingStatus({ tunnelMode: tunnelArg.mode, tunnelUrl: httpUrl, attempt, maxAttempts }))

// Test — was a local copy of the builder, now:
-function buildTunnelWarmingBroadcast({ attempt, maxAttempts, tunnelMode, tunnelUrl }) { /* local duplicate */ }
+import { buildTunnelWarmingStatus, buildTunnelReadyStatus } from '../src/server-cli.js'
```

18/18 tunnel tests pass including the new field-whitelist assertion.

File: packages/desktop/src-tauri/src/server.rs [hunk 4]
```diff
@@ -11,6 +11,75 @@ use crate::config::{self, ChroxyConfig};
...
+    ];
+    let mut out = Vec::new();
+    for (pid, cmd) in procs {
+        // Require the cloudflared binary name as a whole word. We match
+        // on both bare "cloudflared" and full-path invocations like
+        // "/opt/homebrew/bin/cloudflared".
+        let has_cloudflared_binary = cmd.split_whitespace().any(|tok| {
+            let base = tok.rsplit('/').next().unwrap_or(tok);
+            base == "cloudflared"
+        });
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/desktop/src-tauri/src/server.rs:64]: **FIX**

Fixed in `8bff12bd2`

**Change:** Made binary detection cross-platform. The tokenizer now:
- strips surrounding single/double quotes
- takes the basename via splitting on both `/` and `\`
- strips a case-insensitive `.exe` suffix
- compares case-insensitively

Added 4 Windows-focused unit tests locking this in:

```diff
         let has_cloudflared_binary = cmd.split_whitespace().any(|tok| {
-            let base = tok.rsplit('/').next().unwrap_or(tok);
-            base == "cloudflared"
+            let trimmed = tok.trim_matches(|c: char| c == '"' || c == '\'');
+            let base = trimmed
+                .rsplit(|c: char| c == '/' || c == '\\')
+                .next()
+                .unwrap_or(trimmed);
+            let without_exe = if base.len() >= 4
+                && base[base.len() - 4..].eq_ignore_ascii_case(".exe")
+            {
+                &base[..base.len() - 4]
+            } else {
+                base
+            };
+            without_exe.eq_ignore_ascii_case("cloudflared")
         });
```

New tests: `cloudflared_filter_matches_windows_exe` (bare `cloudflared.exe` with backslash path), `cloudflared_filter_matches_quoted_windows_path` (`"C:\\Program Files\\..."` form), `cloudflared_filter_is_case_insensitive_for_binary` (`CLOUDFLARED.EXE`), and `cloudflared_filter_rejects_non_cloudflared_exe` (`someapp.exe` must not match just because it has the URL in argv).

Test count: 14/14 cloudflared tests pass (was 10). Full `cargo test`: 79/79 pass (was 75).

Note: the Windows `kill_orphan_cloudflared` enumeration path still uses `wmic` which is deprecated on newer Windows. That's tracked separately in https://github.com/blamechris/chroxy/issues/2850 since it's a much larger (~100-line) patch.

**Comment by blamechris** [comment]
## Review Comments Addressed

| # | Comment | Outcome | Reference |
|---|---------|---------|-----------|
| 1 | Docstring said `http(s)://` but filter only searches `http://` | FIX | `8bff12bd2` |
| 2 | Windows `wmic` is deprecated — enumeration becomes a no-op on newer Windows | FOLLOW-UP | [#2850](https://github.com/blamechris/chroxy/issues/2850) |
| 3 | Tests claim to pin broadcast payload shape but only exercise `waitForTunnel`/`onAttempt` | FIX | `8bff12bd2` |
| 4 | Binary detection fails on Windows `cloudflared.exe` + backslash/quoted paths | FIX | `8bff12bd2` |

**Total:** 4 comments addressed
- Fixed: 3 (all in commit `8bff12bd2`)
- False positives: 0
- Follow-up issues created: 1 ([#2850](https://github.com/blamechris/chroxy/issues/2850) — Windows `wmic` → PowerShell fallback, ~100 lines of Windows-specific code outside this PR's scope)
- Existing issues closed: 0

### Test delta after fixes
- Rust: **79/79** pass (was 75/75 — added 4 Windows binary detection tests).
- Server: **18/18** tunnel tests pass (was 15/15 — restructured `tunnel-warming-broadcast.test.js` into 7 tests exercising real production helpers).
- Server lint: 0 errors (unchanged warnings).

PR ready for re-review.

**Comment by blamechris** [comment]
## Independent Code Review (second pass): PR #2845

### Summary
Second-pass independent review after the earlier agent-review + check-pr cycle. Reread the full diff and ran a fresh pass against the critical surfaces: the cloudflared filter, tunnel_warming protocol contract, log buffer safety, and the `get_startup_logs` Tauri command.

### Strengths
- `cloudflared_pids_to_kill` is a pure function, extensively tested, and its port/binary discrimination covers the important adversarial cases: substring port collisions (876 vs 8765), cross-port Chroxy instances (explicit test), `tcp://` / non-HTTP schemes, Windows `.exe`/quoted paths, and mixed process listings.
- Production helpers `buildTunnelWarmingStatus` / `buildTunnelReadyStatus` are imported by the tests — the wire shape is now pinned at source, and a whitelist assertion prevents stray-field drift.
- Dual-name handling (`tunnel_warming` + legacy `tunnel_verifying`) is applied consistently across `message-handler.ts`, `FooterBar.tsx`, `ConsolePage.tsx`, `App.tsx`, and `types.ts` — no place reads the raw phase name without normalizing.
- Health-poll logging is now per-attempt with elapsed ms, plus a summary counter on timeout. That alone makes #2832 reproducible from a bug report.

### Issues Found

#### Critical (Must Fix)
None.

#### Suggestions (Should Consider)
None that aren't already tracked. The Windows `wmic` fallback is parked in #2850 with a reasonable rationale.

#### Nitpicks (Optional)
- `buildTunnelWarmingStatus` accepts `attempt` OR `maxAttempts` individually but only emits counters when BOTH are numbers. Passing one without the other silently drops the counter; not harmful since the two always travel together in practice, but a small `assert` or typedef comment would make that invariant explicit.

### Backward-compat note
Newer server → older dashboard: the old handler only recognizes `tunnel_verifying`, so a dashboard pinned to an older release will ignore `tunnel_warming` and show no banner (the transitional "ready" handler still works, so it won't get stuck). Dashboard and server ship together in the Tauri bundle, so this only matters if a user runs a new server against a stale browser tab from a prior bundle — recoverable via refresh.

### Security
- `get_startup_logs` is a Tauri IPC command — no network exposure.
- Log buffer captures stdout/stderr. `server-cli.js` masks the API token by default; the full token only leaks into stdout when `--show-token` is set (opt-in). Acceptable.

### Architecture Notes
The extraction of the broadcast helpers into `server-cli.js` is a small but meaningful improvement — previously the tests constructed the payload themselves and silently accepted drift. Pure `cloudflared_pids_to_kill` with parser adapters per platform (`parse_ps_line` on unix, inline wmic CSV on windows) is the right separation.

### Verdict
- [x] Approve — Ready to merge
- [ ] Request Changes
- [ ] Comment

CI: 12/12 green. No unresolved review threads. No blocking issues found on this second pass.
